File: lib/domain/threats/threat.dart
import 'dart:math' as math;
import '../../core/engine/vec2.dart';

/// What kind of object is converging on the deity at the centre of the arena.
enum ThreatKind {
  /// Common hurled boulder — parry in its direction.
  boulder,

  /// Fast dark bolt — shorter reaction time.
  darkBolt,

  /// Drifting shade — weaves slightly as it approaches.
  shade,

  /// Titan attack — requires several parries before it is repelled.
  titan,

  /// A blessing — should be allowed to reach the centre (do NOT parry).
  blessing,

  /// A coin-like essence mote — also collected by letting it reach the centre.
  essenceMote,
}

/// Lifecycle phase of a [Threat].
enum ThreatPhase { incoming, repelled, collected, expired }

/// A single object travelling along a radius toward the arena core.
///
/// Position is polar ([bearing], [radius]) around the arena centre; the
/// renderer converts it to a screen point. Threats start beyond the arena edge
/// and advance toward the core radius.
class Threat {
  final ThreatKind kind;

  /// Direction from the centre, in radians (screen space, +y down).
  double bearing;

  /// Distance from the centre in logical pixels (decreases over time).
  double radius;

  /// Inward speed in pixels/second.
  double speed;

  /// Sideways weave amount (used by [ThreatKind.shade]).
  final double weaveAmp;
  final double weavePhaseSpeed;
  double _weaveT = 0;

  /// Remaining parries needed (titans need several; others need one).
  int hp;

  /// Visual radius in pixels.
  final double drawRadius;

  ThreatPhase phase = ThreatPhase.incoming;

  /// Animation timer used while playing out repelled/collected/expired.
  double exitT = 0;

  /// Outward velocity used after a repel, so it flies back the way it came.
  double exitSpeed = 0;

  Threat({
    required this.kind,
    required this.bearing,
    required this.radius,
    required this.speed,
    this.weaveAmp = 0,
    this.weavePhaseSpeed = 0,
    this.hp = 1,
    this.drawRadius = 26,
  });

  bool get isBlessing =>
      kind == ThreatKind.blessing || kind == ThreatKind.essenceMote;

  bool get isAlive => phase == ThreatPhase.incoming;

  /// Effective bearing including the shade's gentle weave.
  double get displayBearing =>
      bearing + weaveAmp * 0.06 * math.sin(_weaveT * weavePhaseSpeed);

  /// Advances the threat one step.
  void update(double dt) {
    _weaveT += dt;
    switch (phase) {
      case ThreatPhase.incoming:
        radius -= speed * dt;
        break;
      case ThreatPhase.repelled:
        radius += exitSpeed * dt;
        exitT += dt;
        break;
      case ThreatPhase.collected:
      case ThreatPhase.expired:
        exitT += dt;
        break;
    }
  }

  /// Time (seconds) until this threat reaches [coreRadius] at current speed.
  double timeToCore(double coreRadius) {
    if (speed <= 0) return double.infinity;
    return (radius - coreRadius) / speed;
  }

  /// Screen position given the arena [centre] (uses [displayBearing]).
  Vec2 positionFrom(Vec2 centre) => centre + Vec2.angle(displayBearing, radius);

  void repel() {
    phase = ThreatPhase.repelled;
    exitT = 0;
    exitSpeed = speed * 2.4 + 220;
  }

  void collect() {
    phase = ThreatPhase.collected;
    exitT = 0;
  }

  void expire() {
    phase = ThreatPhase.expired;
    exitT = 0;
  }
}

File: ZeusBoltDash/Domain/Threats/Threat.swift
import Foundation

enum ThreatKind: String, CaseIterable {
    case boulder
    case darkBolt
    case shade
    case titan
    case blessing
    case essenceMote
}

final class Threat: Identifiable {
    let id = UUID()
    let kind: ThreatKind
    var bearing: Double
    var radius: Double
    var speed: Double
    var hp: Int
    var isRepelled = false
    var repelBearing: Double = 0
    var repelSpeed: Double = 0
    /// Pickups are absorbed (not repelled): they pop and fade at the core.
    var isCollected = false
    var collectT: Double = 0

    var sprite: String {
        switch kind {
        case .boulder:     return "boulder_a"
        case .darkBolt:    return "bolt_frame1"
        case .shade:       return "bolt_frame2"
        case .titan:       return "boulder_e"
        case .blessing:    return "essence_a"
        case .essenceMote: return "essence_b"
        }
    }

    var isPickup: Bool { kind == .blessing || kind == .essenceMote }
    var isAlive: Bool { !isRepelled && !isCollected && radius > 0 }

    var position: Vec2 { Vec2(angle: bearing, radius: radius) }

    init(kind: ThreatKind, bearing: Double, radius: Double, speed: Double) {
        self.kind = kind
        self.bearing = bearing
        self.radius = radius
        self.speed = speed
        self.hp = kind == .titan ? 3 : 1
    }

    func update(dt: Double) {
        if isRepelled {
            radius += repelSpeed * dt
            bearing += repelBearing * dt
        } else if isCollected {
            collectT += dt
        } else {
            radius -= speed * dt
        }
    }

    func repel(arena: Double) {
        isRepelled = true
        repelSpeed = speed * 2.2
        repelBearing = Double.random(in: -0.4...0.4)
    }

    /// Absorb this pickup at the core (pop + fade instead of flying away).
    func collect() {
        isCollected = true
        collectT = 0
    }

    var timeToCore: Double {
        guard speed > 0 else { return .infinity }
        return radius / speed
    }
}
